src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -0,0 +1,168 @@
+            }
+
+            AddReceivers();
+
+            foreach (var receiver in receivers)
+            {
+                try
+                {
+                    await receiver.Init().ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Fatal($"Receiver {receiver.Id} failed to initialize.", ex);
+                    throw;
+                }
+            }
+        }
+
+        public void Start()
+        {
+            foreach (var receiver in receivers)
+            {
+                try
+                {
+                    receiver.Start();
+                }
+                catch (Exception ex)
+                {
+                    Logger.Fatal($"Receiver {receiver.Id} failed to start.", ex);
+                    throw;
+                }
+            }
+        }
+
+        public Task Stop()
+        {
+            var receiverStopTasks = receivers.Select(async receiver =>
+            {
+                Logger.DebugFormat("Stopping {0} receiver", receiver.Id);
+                await receiver.Stop().ConfigureAwait(false);
+                Logger.DebugFormat("Stopped {0} receiver", receiver.Id);
+            });
+
+            return Task.WhenAll(receiverStopTasks);
+        }
+
+        public Task CreateQueuesIfNecessary(string username)
+        {
+            if (!IsEnabled)
+            {
+                return TaskEx.CompletedTask;
+            }
+
+            var queueCreator = receiveInfrastructure.QueueCreatorFactory();
+
+            return queueCreator.CreateQueueIfNecessary(queueBindings, username);
+        }
+
+        public async Task PerformPreStartupChecks()
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            var result = await receiveInfrastructure.PreStartupCheck().ConfigureAwait(false);
+
+            if (!result.Succeeded)
+            {
+                throw new Exception($"Pre start-up check failed: {result.ErrorMessage}");
+            }
+        }
+
+        bool IsEnabled => configuration != null;
+
+        void AddReceivers()
+        {
+            var requiredTransactionSupport = configuration.TransactionMode;
+            var recoverabilityExecutorFactory = builder.Build<RecoverabilityExecutorFactory>();
+
+            var recoverabilityExecutor = recoverabilityExecutorFactory.CreateDefault(eventAggregator, configuration.LocalAddress);
+            var pushSettings = new PushSettings(configuration.LocalAddress, errorQueue, configuration.PurgeOnStartup, requiredTransactionSupport);
+            var dequeueLimitations = configuration.PushRuntimeSettings;
+
+            receivers.Add(new TransportReceiver(MainReceiverId, BuildMessagePump(), pushSettings, dequeueLimitations, mainPipelineExecutor, recoverabilityExecutor, criticalError));
+            queueBindings.BindReceiving(configuration.LocalAddress);
```

Comment:
Adding the queues to `queueBindings` used to be done in the `Receiving` feature setup. Since they are here now instead, that means that queue creation is broken because `CreateQueuesIfNecessary` is called before any bindings have been added.

Reply:
I've added tests and a fix for this